src/bento/components/bento-autocomplete/1.0/component.tsx

```diff
@@ -0,0 +1,521 @@
+
+import fuzzysearch from '#third_party/fuzzysearch';
+
+import {AutocompleteItem} from './autocomplete-item';
+import {useStyles} from './component.jss';
+import {
+  getEnabledResults,
+  getItemElement,
+  getSelectedObjectValue,
+  getSelectedTextValue,
+} from './helpers';
+import {HighlightedText} from './highlighted-text';
+import {tokenPrefixMatch} from './token-prefix-match';
+import {
+  BentoAutocompleteProps,
+  InputElement,
+  Item,
+  ItemTemplateFn,
+  OnSelectData,
+  isValidFilterType,
+} from './types';
+import {useAutocompleteBinding} from './use-autocomplete-binding';
+
+// @ts-ignore
+import {addParamToUrl} from '../../../../url';
+
+const INITIAL_ACTIVE_INDEX = -1;
+
+const getItems = (response: {items: Item[]}) =>
+  getValueForExpr(response, 'items');
+
+export function BentoAutocomplete({
+  children,
+  filter = 'none',
+  filterValue = 'value',
+  highlightUserEntry = false,
+  id,
+  inline: inlineTrigger,
+  items = [],
+  itemTemplate,
+  maxItems,
+  minChars = 1,
+  onError = () => {},
+  onSelect = () => {},
+  parseJson = getItems,
+  prefetch = false,
+  src,
+  submitOnEnter = false,
+  suggestFirst = false,
+  query,
+}: BentoAutocompleteProps) {
+  const elementRef = useRef<HTMLElement>(null);
+  const containerId = useValueRef<string>(
+    id || `${Math.floor(Math.random() * 100)}_AMP_content_`
+  );
+  const inputRef = useRef<HTMLInputElement>(null);
+  const containerRef = useRef<HTMLDivElement>(null);
+
+  const [substring, setSubstring] = useState<string>('');
+  const [activeIndex, setActiveIndex] = useState<number>(INITIAL_ACTIVE_INDEX);
+  const [areResultsDisplayed, setAreResultsDisplayed] =
+    useState<boolean>(false);
+  const [shouldSuggestFirst, setShouldSuggestFirst] =
+    useState<boolean>(suggestFirst);
+  const classes = useStyles();
+  const [hasFetchedInitialData, setHasFetchedInitialData] = useState(false);
+  // In order to enable refetching, we currently have to disable the request between
+  // and re-enable it on user input.
+  const [hasFetchedData, setHasFetchedData] = useState(false);
+
+  const binding = useAutocompleteBinding(inlineTrigger);
+
+  const url = useMemo(() => {
+    if (src && query) {
+      // Don't create an invalid URL if there is a query key but not a value
+      return substring ? addParamToUrl(src, query, substring) : null;
+    }
+    return src;
+  }, [src, substring, query]);
+
+  const isFetchEnabled = useMemo(() => {
+    if (url) {
+      if (query) {
+        return !hasFetchedData;
+      }
+      if (!hasFetchedInitialData) {
+        return prefetch || areResultsDisplayed;
+      }
+    }
+    return false;
+  }, [
+    url,
+    hasFetchedInitialData,
+    hasFetchedData,
+    query,
+    prefetch,
+    areResultsDisplayed,
+  ]);
+
+  const shouldRefetch = useMemo(() => {
+    return !!query;
+  }, [query]);
+
+  const {data} = useQuery<Item[]>(
+    () => {
+      return xhrUtils.fetchJson(url!).then((response) => parseJson(response));
```

Comment:
Originally I had `fetchJson` included as a prop in the component, but we are stubbing `xhrUtils` in the tests anyway, so including it as a dependency rather than directly calling `xhrUtils.fetchJson` didn't do a whole lot of good here.